Generalize sstable::move_to_new_dir() method

xemul opened the pull request:
This method requires callers to remember that the sstable is the collection of files on a filesystem and to know what exact directory they are all in. That's not going to work for object storage, instead, sstable should be moved between more abstract states.

This PR replaces move_to_new_dir() call with the change_state() one that accepts target sub-directory string and moves files around. Currently supported state changes:

* staging -> normal
* upload -> normal | staging
* any -> quarantine

All are pretty straightforward and move files between table basedir subdirectories with the exception that upload -> quarantine should move into upload/quarantine subdirectory. Another thing to keep in mind, that normal state doesn't have its subdir but maps directory to table's base directory.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4024/

denesb commented:
> This PR breaks the move_to_new_dir() call into two reflecting two existing sstable state changes:

Which is it? reflecting or existing?

denesb commented on sstables/sstables.hh:
```diff
@@ -498,6 +504,7 @@ public:
         future<> snapshot(const sstable& sst, sstring dir, absolute_path abs) const;
         future<> quarantine(const sstable& sst, delayed_commit_changes* delay);
         future<> activate(const sstable& sst, generation_type gen, delayed_commit_changes* delay);
+        future<> to_staging(const sstable& sst, generation_type gen);
```
The other names are verbs, while this is just a direction.
Not sure how  verb version of this would sound like though.

xemul replied:
> Not sure how verb version of this would sound like though.

That's exactly the reason it's named like that. 
Would put_on_stage() be too dramatic?

bhalevy commented on sstables/sstables.cc:
```diff
@@ -2230,6 +2230,18 @@ future<> sstable::move_to_new_dir(sstring new_dir, generation_type new_generatio
     _generation = std::move(new_generation);
 }
 
+future<> sstable::filesystem_storage::activate(const sstable& sst, generation_type gen, delayed_commit_changes* delay_commit) {
```
But the sstable is acrive also when its in staging.
While it's there it is not subject to compaction but otherwise it participates in queries etc.

xemul replied:
Well, it depends on what "active" mean :) I haven't met this term in sstables.cc file, so I didn't suggest that active sstable would really mean something well defined.

Would you suggest the better name, please?

raphaelsc replied:
I also had an issue with "activate" for the same reason. quarantine, upload, staging are all conditions. a sstable no in those conditions is a normal one. normal = "characterized by that which is considered usual, typical, or routine". so the verb can be **normalize** which means "bring or return to a normal or standard condition or state."

fits very well

xemul replied:
I'm OK with "normalize" (though my linear algebra professor managed to create incredibly strong "normal == perpendicular" association)

bhalevy commented on sstables/sstables.cc:
```diff
@@ -2242,6 +2246,19 @@ future<> sstable::move_from_staging(delayed_commit_changes* delay_commit) {
     return _storage.activate(*this, generation(), delay_commit);
 }
 
+future<> sstable::move_from_upload(generation_type new_generation, to_staging staging) {
```
why have a function for every source location and every destination location?
how about a generic move or relocate function that gets the destination location (and maybe the source location if there are source-dependent side effects that need to take place. this can also be implemented with a given functor passed by the caller)

xemul replied:
There is one already, but on the storage level. For sstable level, currently the state changes are:
* any -> quarantine
* upload -> base
* upload -> staging
* staging -> base

And not all any -> any are meaningful, e.g. what would quarantine -> upload mean? I don't know :)

Having a generic function that can do any -> any would require it to validate that the transition is "supported", but what's the point? Also, some transitions require generation argument, so there will also need to be some "checks" for the provided generation to be OK.

xemul replied:
Re-implemented it this way. Target directory selection becomes ... pretty hairy :)

xemul commented:
> > This PR breaks the move_to_new_dir() call into two reflecting two existing sstable state changes:
> 
> Which is it? reflecting or existing?

Erm... I don't follow :) There are two places where the move_to_new_dir() is called -- to make `staging -> base` and `upload -> base | staging` moves. So I replace the move_to_new_dir() with `move_from_staging` and `move_from_upload`

xemul commented:
rebased (had conflicts)

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4062/

raphaelsc commented on replica/distributed_loader.cc:
```diff
@@ -267,14 +267,18 @@ distributed_loader::reshape(sharded<sstables::sstable_directory>& dir, sharded<r
 distributed_loader::make_sstables_available(sstables::sstable_directory& dir, sharded<replica::database>& db,
-        sharded<db::view::view_update_generator>& view_update_generator, fs::path datadir, sstring ks, sstring cf) {
+        sharded<db::view::view_update_generator>& view_update_generator, bool to_staging, sstring ks, sstring cf) {
 
```
perhaps rename to_staging to needs_view_update, so staging knowledge is confined to make_sstables_available().

raphaelsc replied:
staging is an implementation detail of persisting knowledge of sstables needing view update.

xemul commented:
updated:
* renamed all storage methods to be "directions"
* renamed active to normal
* renamed dist-loader's to_staging variable to needs_view_update
* made storage::move() method private

avikivity commented on sstables/sstables.cc:
```diff
@@ -2230,6 +2230,18 @@ future<> sstable::move_to_new_dir(sstring new_dir, generation_type new_generatio
+    }
+
+    return _storage.to_normal(*this, generation(), delay_commit);
+}
```
This is inversion of roles. Instead of sstable being a building block and the control plane deciding where to put it, now the sstable knows about view building and the staging directory.

xemul replied:
Agree. The move-to-state call should work better.

xemul commented on sstables/sstables.hh:
```diff
@@ -497,6 +503,7 @@ public:
         future<> seal(const sstable& sst);
         future<> snapshot(const sstable& sst, sstring dir, absolute_path abs) const;
         future<> to_quarantine(const sstable& sst, delayed_commit_changes* delay);
+        future<> to_staging(const sstable& sst, generation_type gen);
```
But that's the whole point of sstable::storage (well, currently it's filesystem_storage, but #12523 generalizes it) to "know" about directories. Caller knows sstable can be in one of pre-defined "states" -- normal, staging, quarantine, upload (etc?), passes this "state" down to sstables (my next PR is going to be about it) then it's up to storage driver to decide how to implement this "state"

avikivity replied:
I think it should be split out. The part that stays in sstables/ shouldn't know about staging/update/etc. The part that moves to replica/ can.

Perhaps sstables::strorage should represent a single directory, and we'll have sstables::sstable::move_to(storage&). Only compatible moves will be supported. We'll have `sstables::storage _upload_dir` in `replica::table`.

xemul replied:
> Perhaps sstables::strorage should represent a single directory

After more thinking I disagree. The /var/lib/something/staging directory describes both -- location and state (or status) of an sstable. Thus having different sstable::storage instances to describe different _states_ is not correct. State should rather be the member of the class sstable itself and the storage should just point to the sstable location. In particular, filesystem_storage should have `dir` point to /var/lib/something, not to /var/lib/something/staging. With that when asked to do something with an sstable the filesystem storage driver should compose the path like `path = this->dir + '/' + sstable->state + '/' sstable->filename(component)`, and s3 driver would just stick to its bucket and the `location` column of the ownership table.

xemul commented:
updated:
* keep single `change_state()` method that accepts target subdirectory name, not the full path, and moves files around

raphaelsc commented on sstables/sstables.hh:
```diff
@@ -221,6 +222,12 @@ public:
+    // Known states are normal, staging, upload and quarantine.
+    // It's up to the storage driver how to implement this.
+    future<> change_state(sstring to, generation_type new_generation, delayed_commit_changes* delay = nullptr);
+
```
we rely on stable generations in the sstable set reader. we don't have users changing sstable generation from X to Y. so we can kill new_generation param. we don't need to do it in this patch though.

xemul replied:
> we don't have users changing sstable generation from X to Y

Not quite

```c++
distributed_loader::make_sstables_available() {
    ...
        auto gen = table.calculate_generation_for_new_table();
        co_await sst->change_state(state, gen);
    ...
```

Previous version mitigated this by introducing specific methods to change states, in particular there was `move_from_upload` method that accepted target generation parameter

raphaelsc replied:
you're right. that's only needed when moving from upload to avoid generation clash. less of a problem with UUID. although user could mistankely upload the same sstable more than once, and that should still work.

as the change state interface is now generic in the latest version, it becomes hard to put contraints on specific state switches. if we're moving a sstable in-use (e.g. in base directory), changing generation should be disallowed.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4107/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4120/

xemul commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4120/

```
[2023-02-01T16:06:15.755Z] ERROR 2023-02-01 18:06:10,833 [shard 0] view - Error applying view update to 127.56.208.5 (view: cql_test_1675267516773.cql_test_1675267569717, base token: -7160136740246525330, view token: -7160136740246525330): data_dictionary::no_such_column_family (Can't find a column family with UUID 577a5cf0-a24a-11ed-99d7-589fb1ba5fe5)
...
[2023-02-01T16:06:15.769Z] ERROR 2023-02-01 18:06:12,217 [shard 0] mutation_reader - [validator 0x61600088eb10 for mutation_compactor for read (cql_test_1675267516773.cql_test_1675267517152 555f9340-a24a-11ed-99d7-589fb1ba5fe5)] Unexpected mutation fragment: previous partition start, current partition start, at: 0xe3798d0 0x14ae6415 0x14ae616d 0x14ae8583 0x14ae900f 0x13b2251d 0x13b21e14 0xf6af066 0xf6b0641 0xf029ec7 0x11369219 0x11368f4e 0x11368cf9 0x11368cb7 0x11368c34 0x113688ab 0x1136832e 0x113680a3 0x1135c113 0x11303786 0x11347e97 0x1418c6a9
```

:open_mouth:

xemul commented:
Local run passed :confused: re-kicked the CI, let's see

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4123/

xemul commented:
> > CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4120/
> 
> ```
> [2023-02-01T16:06:15.755Z] ERROR 2023-02-01 18:06:10,833 [shard 0] view - Error applying view update to 127.56.208.5 (view: cql_test_1675267516773.cql_test_1675267569717, base token: -7160136740246525330, view token: -7160136740246525330): data_dictionary::no_such_column_family (Can't find a column family with UUID 577a5cf0-a24a-11ed-99d7-589fb1ba5fe5)
> ...
> [2023-02-01T16:06:15.769Z] ERROR 2023-02-01 18:06:12,217 [shard 0] mutation_reader - [validator 0x61600088eb10 for mutation_compactor for read (cql_test_1675267516773.cql_test_1675267517152 555f9340-a24a-11ed-99d7-589fb1ba5fe5)] Unexpected mutation fragment: previous partition start, current partition start, at: 0xe3798d0 0x14ae6415 0x14ae616d 0x14ae8583 0x14ae900f 0x13b2251d 0x13b21e14 0xf6af066 0xf6b0641 0xf029ec7 0x11369219 0x11368f4e 0x11368cf9 0x11368cb7 0x11368c34 0x113688ab 0x1136832e 0x113680a3 0x1135c113 0x11303786 0x11347e97 0x1418c6a9
> ```
> 
> open_mouth (edit: `cql-pytest debug [ FAIL ] cql-pytest.test_materialized_view.1 56.79s`)

@denesb , does this :point_up: ring any bell?

denesb commented:
> > > CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4120/
> > 
> > 
> > ```
> > [2023-02-01T16:06:15.755Z] ERROR 2023-02-01 18:06:10,833 [shard 0] view - Error applying view update to 127.56.208.5 (view: cql_test_1675267516773.cql_test_1675267569717, base token: -7160136740246525330, view token: -7160136740246525330): data_dictionary::no_such_column_family (Can't find a column family with UUID 577a5cf0-a24a-11ed-99d7-589fb1ba5fe5)
> > ...
> > [2023-02-01T16:06:15.769Z] ERROR 2023-02-01 18:06:12,217 [shard 0] mutation_reader - [validator 0x61600088eb10 for mutation_compactor for read (cql_test_1675267516773.cql_test_1675267517152 555f9340-a24a-11ed-99d7-589fb1ba5fe5)] Unexpected mutation fragment: previous partition start, current partition start, at: 0xe3798d0 0x14ae6415 0x14ae616d 0x14ae8583 0x14ae900f 0x13b2251d 0x13b21e14 0xf6af066 0xf6b0641 0xf029ec7 0x11369219 0x11368f4e 0x11368cf9 0x11368cb7 0x11368c34 0x113688ab 0x1136832e 0x113680a3 0x1135c113 0x11303786 0x11347e97 0x1418c6a9
> > ```
> >   
> > open_mouth (edit: `cql-pytest debug [ FAIL ] cql-pytest.test_materialized_view.1 56.79s`)
> 
> @denesb , does this point_up ring any bell?

@nyh opened an issue for this, it is seen from time-to-time in CI.

xemul commented:
updated:
- rebased (had conflicts)
- detach `pick_up_from_upload()` from the new `change_state()` one. The latter works for object-storage branch, but the former doesn't, so keep it aside for a while

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4543/

denesb commented on sstables/sstables.hh:
```diff
@@ -220,6 +221,12 @@ public:
+    //
+    // Known states are normal, staging, upload and quarantine.
+    // It's up to the storage driver how to implement this.
+    future<> change_state(sstring to, delayed_commit_changes* delay = nullptr);
```
Would love for `to` to be an `enum class` instead of a string.

xemul replied:
Absolutely. Once I finish #12707 , it _will_ be an `enum class`
